AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -796,17 +798,142 @@ - (BOOL)shouldCalculateLayoutWithConstrainedSize:(ASSizeRange)constrainedSize pa
+  [self __setNeedsLayout];
 }
 
-- (BOOL)canLayoutAsynchronous
+/**
+ * Invalidate the current calculated layout so it can be recalculated in the next layout pass
+ */
+- (void)__setNeedsLayout
 {
-  return !self.isNodeLoaded;
+  ASDN::MutexLocker l(__instanceLock__);
+
+  _calculatedDisplayNodeLayout->invalidate();
+  _pendingDisplayNodeLayout->invalidate();
 }
 
-#pragma mark - Automatic Hierarchy
+/**
+ * Returns a BOOL indicating whether the layer has been marked as needing a layout update.
+ */
+- (BOOL)__needsLayout
+{
+  return _calculatedDisplayNodeLayout->isDirty();
+}
+
+/**
+ * The node's supernodes are traversed until a ancestor node is found that does not require layout. Then layout
+ * is performed on the entire node-tree beneath that ancestor
+ */
+- (void)__layoutIfNeeded
+{
+  ASDisplayNodeAssertThreadAffinity(self);
+  
+  __instanceLock__.lock();
+  
+  // Invalidate the current calculated layout so in the next layout pass it get's recalculated
+  [self invalidateCalculatedLayout];
+  
+  // Walk the node's supernodes until a ancestor node is found that does not require layout
+  if ([_supernode __needsLayout]) {
+    ASDisplayNode *supernode = _supernode;
+    __instanceLock__.unlock();
+    [_supernode __layoutIfNeeded];
+    return;
+  }
+
+  // Layout all subviews starting from the first node that needs layout
+  [self __layout];
+  
+  __instanceLock__.unlock();
+}
+
+/**
+ * Called from [CALayer layoutSublayers:]: All subnode layout related code should happen in there
+ */
+- (void)__layout
+{
+  ASDisplayNodeAssertMainThread();
+  ASDN::MutexLocker l(__instanceLock__);
+  CGRect bounds = _threadSafeBounds;
+  
+  // Prevent zero bounds subnode adjustments
+  if (CGRectEqualToRect(bounds, CGRectZero)) {
+    // Performing layout on a zero-bounds view often results in frame calculations with negative sizes after applying
+    // margins, which will cause layoutThatFits: on subnodes to assert.
+    LOG(@"Warning: No size given for node before node was trying to layout itself: %@. Please provide a frame for the node.", self);
+    return;
+  }
+
+  // Node needs to be measured in case the bounds changed from the current calculated layout
+  [self calculateLayoutIfNecessaryWithBounds:bounds];
+  
+  // Handle placeholder layer creation in case the size of the node changed after the initial placeholder layer
+  // was created
+  if ([self _shouldHavePlaceholderLayer]) {
+    [self _setupPlaceholderLayerIfNeeded];
+  }
+  _placeholderLayer.frame = bounds;
+  
+  // Layout all subnodes based on the layout determined before
+  [self layout];
+  [self layoutDidFinish];
+}
+
+/*
+ * Executes a measure pass if necessary based on the current bounds as constrained size
+ */
+- (void)calculateLayoutIfNecessaryWithBounds:(CGRect)bounds
+{
+  // Setup new constrained and parent size based on bounds
+  ASSizeRange constrainedSize = ASSizeRangeMake(bounds.size);
+  CGSize parentSize = constrainedSize.max;
```

Comment:
I think this would read a bit more logically as:
parentSize = bounds.size;
constrainedSize = SizeMake(parentSize)

Rather than unpacking from the SizeRange to get the input value back.